src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -168,6 +266,259 @@ private static IDataTransform Create(IHostEnvironment env, Arguments args, IData
+                if (Session.Graph[args.LearningRateOperation] == null)
+                    throw _host.ExceptParam(nameof(args.LearningRateOperation), $"'{args.LearningRateOperation}' does not exist in the model");
+            }
+        }
+
+        private void TrainCore(TrainingArguments args, string model, IDataView input)
+        {
+            var inputsForTraining = new string[Inputs.Length + 1];
+            var inputColIndices = new int[inputsForTraining.Length];
+            var isInputVector = new bool[inputsForTraining.Length];
+            var tfInputTypes = new TFDataType[inputsForTraining.Length];
+            var tfInputShapes = new TFShape[inputsForTraining.Length];
+
+            for (int i = 0; i < Inputs.Length; i++)
+            {
+                inputsForTraining[i] = Inputs[i];
+            }
+
+            inputsForTraining[inputsForTraining.Length - 1] = args.LabeLColumn;
+
+            var inputSchema = input.Schema;
+            for (int i = 0; i < inputsForTraining.Length; i++)
+            {
+                if (!inputSchema.TryGetColumnIndex(inputsForTraining[i], out inputColIndices[i]))
+                    throw _host.Except($"Column {inputsForTraining[i]} doesn't exist");
+
+                var type = inputSchema.GetColumnType(inputColIndices[i]);
+                isInputVector[i] = type.IsVector;
+
+                var tfInput = new TFOutput(Graph[inputsForTraining[i]]);
+                tfInputTypes[i] = tfInput.OutputType;
+                tfInputShapes[i] = Graph.GetTensorShape(tfInput);
+                if (tfInputShapes[i].NumDimensions != -1)
+                {
+                    var newShape = new long[tfInputShapes[i].NumDimensions];
+                    newShape[0] = tfInputShapes[i][0] == -1 ? args.BatchSize : tfInputShapes[i][0];
+
+                    for (int j = 1; j < tfInputShapes[i].NumDimensions; j++)
+                        newShape[j] = tfInputShapes[i][j];
+                    tfInputShapes[i] = new TFShape(newShape);
+                }
+
+                var expectedType = TensorFlowUtils.Tf2MlNetType(tfInputTypes[i]);
+                if (type.ItemType != expectedType)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", inputsForTraining[i], expectedType.ToString(), type.ToString());
+            }
+
+            var fetchList = new List<string>();
+            if (args.LossOperation != null)
+                fetchList.Add(args.LossOperation);
+            if (args.MetricOperation != null)
+                fetchList.Add(args.MetricOperation);
+
+            for (int epoch = 0; epoch < args.Epoch; epoch++)
+            {
+                using (var cursor = input.GetRowCursor(a => true))
+                {
+                    var srcTensorGetters = GetTensorValueGetters(cursor, inputColIndices, isInputVector, tfInputTypes, tfInputShapes);
+
+                    float loss=0;
+                    float metric=0;
+                    int rows = 0;
```

Comment:
>int [](start = 20, length = 3)

technically we capable to have more than int.MaxValue rows in dataView, so can you use long here?

Reply:
Yup.

---
In reply to: [221767815](https://github.com/dotnet/machinelearning/pull/1063#discussion_r221767815) [](ancestors = 221767815)